[WIP] Merge feature/garrison into base

zach2good opened the pull request:
<!-- Remove space and place 'x' mark between square [] brackets or click the checkbox after saving to affirm the following points: -->
<!-- (it should look like this: - [x] I have ...) -->
**_I affirm:_**
- [x] I have paid attention to this example and will edit again if need be to not break the formatting, or I will be ignored
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md)
- [x] I've _**tested my code and the things my code has changed**_ since the last commit in the PR, and will test after any later commits

## What does this pull request do?

This is a long-term tracking PR to watch `feature/garrison` and its journey from feature branch + `everything` into `base`.

Previous PRs:
https://github.com/LandSandBoat/server/pull/1796
https://github.com/LandSandBoat/server/pull/429
https://github.com/LandSandBoat/server/pull/424

## Steps to test these changes

Play some Garrison, build a mannequin. etc.

RunAwayDSP committed:
[WIP] Garrison

Co-authored-by: Zach Toogood <zrtoogood@gmail.com>

RunAwayDSP committed:
Update Dropid

merge conflict

more conflict updates

garrison update

RunAwayDSP committed:
North Gust and West Saru

North Gust wont work until all mobs have spawn pos's

RunAwayDSP committed:
Garrison Cleanup

More Code Cleanup

RunAwayDSP committed:
Garrison Spawn Points

North Gustaberg Spawn Points

RunAwayDSP committed:
Garrison Remove Effects on Win

RunAwayDSP committed:
Garrison Jugner Forrest

Add Jugner Forrest Garrison

RunAwayDSP committed:
Garrison Marshlands

Add Garrison Marshlands

RunAwayDSP committed:
Garrison Style Cleanup

Cleanup Tables styling

RunAwayDSP committed:
Garrison BUBURIMU_PENINSULA

Add BUBURIMU_PENINSULA Garrison

RunAwayDSP committed:
Garrison MERIPHATAUD_MOUNTAINS

Add MERIPHATAUD_MOUNTAINS Garrison

RunAwayDSP committed:
Garrison Qufim

Add Qufim Island Garrison

RunAwayDSP committed:
Garrison OnZone Init Create NPC table

Creates NPCs table on server statup/zoneinit stores all the ID's in garrison_data

RunAwayDSP committed:
Garrison Cleanup

Remove Unused Vars

RunAwayDSP committed:
Garrison Add BEAUCEDINE_GLACIER

Add BEAUCEDINE_GLACIER
and Move Create NPC Table to zone tick (zone init appears to be too early)

RunAwayDSP committed:
Garrison THE_SANCTUARY_OF_ZITAH

Add THE_SANCTUARY_OF_ZITAH Garrison

RunAwayDSP committed:
Garrison YUHTUNGA_JUNGLE

Add YUHTUNGA_JUNGLE Garrison

RunAwayDSP committed:
Garrison XARCABARD

Add XARCABARD Garrison

RunAwayDSP committed:
Garrison Add EASTERN_ALTEPA_DESERT

Add EASTERN_ALTEPA_DESERT

RunAwayDSP committed:
Garrison Pathing

Adds Garrison Pathing Thanks to Ophelia

RunAwayDSP committed:
Garrison YHOATOR_JUNGLE

Add YHOATOR_JUNGLE

RunAwayDSP committed:
Garrison Cape Terrigan

Add Cape Terrigan

RunAwayDSP committed:
Garrison Spawn Intervals

Spawn the mobs in intervals during the waves

RunAwayDSP committed:
Garrison NPC Messages and Rank

Add Garrison Messages
Add Rank Requirement

RunAwayDSP committed:
Garrison Fix Bugs

Send NPC back to spawn point
Despawn all npcs before starting incase some are still around.

RunAwayDSP committed:
Garrison Add Random Function

Mobs spawn randomly from available table (matches retail behavior now)

RunAwayDSP committed:
Garrison Update NPC Strength

NPCs were too strong this is more accurate

RunAwayDSP committed:
Garrison Get Functioning Names

Issue with renaming is the packet offset for "equipped" not aligning so had to change to a different look in database, and assign a new look on spawn

RunAwayDSP committed:
Garrison Gil Reward

Add Gil Reward for victory

Garrison Gil/Cleanup

Gil text
Code Cleanup/remove unneeded lines
More cleanup to come

Garrison Text Correction

Fix Outpost Text when starting garrison

zach2good commented on scripts/settings/default/main.lua:
```diff
@@ -199,4 +199,12 @@ xi.settings =
     GOBBIE_BOX_MIN_AGE           = 45,   -- Minimum character age in days before a character can sign up for Gobbie Mystery Box
+
+    -- GARRISON
+    GARRISON_LOCKOUT             = 1800; -- Time in seconds before a new garrison can be started (default: 1800)
```
These lines should end with commas, not `;`'s

zach2good commented on scripts/zones/Beaucedine_Glacier/IDs.lua:
```diff
@@ -29,6 +29,7 @@ zones[xi.zone.BEAUCEDINE_GLACIER] =
         YOU_CANNOT_ENTER_DYNAMIS       = 7882,  -- You cannot enter Dynamis - [Dummy/San d'Oria/Bastok/Windurst/Jeuno/Beaucedine/Xarcabard/Valkurm/Buburimu/Qufim/Tavnazia] for <number> [day/days] (Vana'diel time).
         PLAYERS_HAVE_NOT_REACHED_LEVEL = 7884,  -- Players who have not reached level <number> are prohibited from entering Dynamis.
         DYNA_NPC_DEFAULT_MESSAGE       = 7894,  -- There is an unusual arrangement of branches here.
+        GARRISON                       = 8244,  -- Hm? What is this? How do I know this is not some trick?.
```
This could do with a better name, also, `?.` looks kind of fishy, my extracts list it as:
```
Hm? What is this? %? How do I know this is not some [San d'Orian/Bastokan/Windurstian] trick?
```

(we need this to line up with the extract, for our update tooling to continue to work)

https://raw.githubusercontent.com/zach2good/UpdateExtractor/versions/2021-december/out/scripts/zones/Beaucedine_Glacier/Text.lua.raw

zach2good replied:
Also `GARRISON_BASE` would be an OK name

zach2good commented on scripts/zones/Beaucedine_Glacier/npcs/Parledaire_RK.lua:
```diff
@@ -14,11 +16,48 @@ local guardRegion = xi.region.FAUREGANDI
+    local zoneId = npc:getZoneID()
+    local text = zones[zoneId].text
+    local win = player:getZone():getLocalVar(string.format("[GARRISON]Treasure_%s", zoneId))
+    local won = player:getCharVar("Garrison_Won")
```
Could these be condensed down into a single `GarrisonStatus` var?

RunAwayDSP replied:
needs better naming:
win is when the trader has won the garrison but has yet to trigger the win with the guard.
won is after the trader has triggered the win other players can now talk to npc to remove effects and get the won text.

needs 2 different vars though for the 2 different triggers

RunAwayDSP replied:
fixed in #1919

zach2good commented on scripts/zones/Buburimu_Peninsula/npcs/Bonbavour_RK.lua:
```diff
@@ -14,11 +16,48 @@ local guardRegion = xi.region.KOLSHUSHU
+        xi.garrison.onLose(player, npc)
+    else
+        xi.conq.overseerOnTrigger(player, npc, guardNation, guardType, guardEvent, guardRegion)
+    end
```
This logic looks quite similar between guards, could probably be extracted into a shared function

RunAwayDSP replied:
fixed in #1919

zach2good commented on sql/mob_droplist.sql:
```diff
@@ -17416,29 +17416,31 @@ INSERT INTO `mob_droplist` VALUES (3222,2,0,1000,1455,0); -- one_byne_bill
+INSERT INTO `mob_droplist` VALUES (3222,0,0,1000,16360,10);
+INSERT INTO `mob_droplist` VALUES (3223,0,0,1000,16017,150);
+INSERT INTO `mob_droplist` VALUES (3224,0,0,1000,1604,100); -- Mannequin Legs
+INSERT INTO `mob_droplist` VALUES (3224,0,0,1000,1605,100); -- Mannequin Feet
```
I chewwed up these db changes during my rebasing. They'll need to adopt the new `@COMMON` syntax. Also, these could be PR'd to `base` ahead of time so that they don't conflict with any other db changes.

RunAwayDSP committed:
Garrison Cleanup

Fixed Text ID Renamed to GARRISON_BASE

Fixed Settings converted ; to ,

Fixed droplist with UNCOMMON

zach2good committed:
Merge pull request #1911 from RunAwayDSP/feature/garrison

Garrison Cleanup

RunAwayDSP committed:
Garrison Clean Up and Begin ID Shift Prevention

Cleaned up onTrade moving all the checks into garrison.lua

Added onTrigger moving all checks into garrison.lua

Should no longer need to touch all 48 Outpost Guards

Cleaned up Charvar to use Status instead of won or lost (still need local var for the trader)

Converted many mob id to be shift resistant (not 100% yet)

Added garrison related items to items.lua and converted to using items.lua

Added boss offsets for id shifting (still need to add lookup for boss id)

RunAwayDSP committed:
Garrison Clean Up and Begin ID Shift Prevention

Cleaned up onTrade moving all the checks into garrison.lua

Added onTrigger moving all checks into garrison.lua

Should no longer need to touch all 48 Outpost Guards

Cleaned up Charvars to use Status, only need the one var now

Added garrison related items to items.lua and converted to using items.lua

Added boss offsets for id shifting (still need to add lookup for boss id)

RunAwayDSP committed:
Merge branch 'feature/garrison' of https://github.com/RunAwayDSP/server into feature/garrison

RunAwayDSP committed:
Merge branch 'feature/garrison' of https://github.com/RunAwayDSP/server into feature/garrison

RunAwayDSP committed:
Merge branch 'feature/garrison' of https://github.com/RunAwayDSP/server into feature/garrison

RunAwayDSP committed:
Garrison Name and Look Tables

Update to using name and look tables
Entity Packet Needs update to handle Equipped Look still using model_ids until  fixed

Some of the NPC names still need to be captured  couldnt find any data on those names

Add npc cleanup in cases where they respawn when garrison isnt running

RunAwayDSP committed:
Merge branch 'LandSandBoat:feature/garrison' into feature/garrison

RunAwayDSP committed:
Fix Beastmen Controlled Error

Dont try to spawn npc tables if outpost controlls the zone

zach2good committed:
Merge pull request #1919 from RunAwayDSP/feature/garrison

Garrison Clean Up and ID Shift Prevention

zach2good commented on scripts/zones/Beaucedine_Glacier/npcs/Akane_IM.lua:
```diff
@@ -15,10 +17,12 @@ local guardEvent  = 32761
 
 entity.onTrade = function(player, npc, trade)
     xi.conq.overseerOnTrade(player, npc, trade, guardNation, guardType)
+    xi.garrison.onTrade(player, npc, trade, guardNation)
 end
 
 entity.onTrigger = function(player, npc)
     xi.conq.overseerOnTrigger(player, npc, guardNation, guardType, guardEvent, guardRegion)
+    xi.garrison.onTrigger(player, npc)
 end
```
These sort of things should probably not run immediately after each other, even if the implementations inside these functions are safe to do so.

```lua
if xi.conq.overseerOnTrade(player, npc, trade, guardNation, guardType) then
    return
else if xi.garrison.onTrade(player, npc, trade, guardNation) then
    return
end
```

This guarantees that only one of these runs, never both. Looks like all the guards need this

RunAwayDSP replied:
so the trade items dont overlap I found no issues getting both to work but I will double check this

zach2good replied:
It doesn't matter if they logically won't overlap, at a glance it's impossible to tell if you're a passer-by. These should be split in a way that makes it obvious that only one can be firing at a time (if/else blocks)

zach2good commented on scripts/zones/Buburimu_Peninsula/Zone.lua:
```diff
@@ -27,6 +28,10 @@ zone_object.onInitialize = function(zone)
     xi.helm.initZone(zone, xi.helm.type.LOGGING)
 end
 
+zone_object.onZoneTick = function(zone)
+    xi.garrison.buildNpcTable(zone)
+end
+
```
This looks very suspicious.

Should you be building the table on every tick?

If it's a thing that needs to be built once, it can go in zone init, or whatever.

If it's something that _should_ happen every tick, this should probably indicate that it's only for when garrison is active

```
if xi.garrison.isGarrisonActive(zone) then
    xi.garrison.tickGarrison(zone)
end
```

etc.

RunAwayDSP replied:
not every tick its there for testing for now.. needs to be moved to something like on conquest update and need to run a check before start to possibly build it as well

zach2good reviewed (commented):
Looks like it's getting closer 👍

RunAwayDSP committed:
Move buildNpcTable to onConquestUpdate

More garrison Fixes.. moved when the npcs are built,
No longer need the Trader SQL (uses an existing base npc from Black Coffin)
Cape Terrigan is 99 Restricted not 75 on retail

zach2good committed:
Merge pull request #2084 from RunAwayDSP/feature/garrison

Move buildNpcTable to onConquestUpdate

zach2good commented:
With this: https://github.com/LandSandBoat/server/pull/2106
You should be able to get the NPCs working correctly

zach2good committed:
Merge branch 'base' into feature/garrison

zach2good commented:
I don't know if the special spawning animation is used in Garrison, but its available: https://github.com/LandSandBoat/server/pull/2144

RunAwayDSP commented:
> I don't know if the special spawning animation is used in Garrison, but its available: #2144

There is its on my todo list just been busy

zach2good commented:
The majority of the useful things from this have been ported into base with:

https://github.com/LandSandBoat/server/pull/2673
https://github.com/LandSandBoat/server/pull/2675
https://github.com/LandSandBoat/server/pull/2681
https://github.com/LandSandBoat/server/pull/2853
https://github.com/LandSandBoat/server/pull/2684
https://github.com/LandSandBoat/server/pull/2203
https://github.com/LandSandBoat/server/pull/2144
https://github.com/LandSandBoat/server/pull/2953
